src/Stratis.Bitcoin/P2P/PeerSelector.cs

```diff
@@ -30,70 +30,99 @@ public interface IPeerSelector
         IEnumerable<PeerAddress> SelectPeers();
     }
 
-    /// <summary>
-    /// Class that solely deals with providing the peer connectors a peer to connect to.
-    /// </summary>
     public sealed class PeerSelector : IPeerSelector
     {
         /// <summary>
         /// The address manager instance that holds the peer list to be queried. 
         /// </summary>
-        private readonly IPeerAddressManager peerAddressManager;
+        private readonly ConcurrentDictionary<IPEndPoint, PeerAddress> peerAddresses;
 
         /// <summary>
         /// Constructor for the peer selector.
         /// </summary>
-        /// <param name="peerAddressManager">The singleton peer address manager instance.</param>
-        public PeerSelector(IPeerAddressManager peerAddressManager)
+        /// <param name="peerAddresses">The collection of peer address as managed by the peer address manager.</param>
+        public PeerSelector(ConcurrentDictionary<IPEndPoint, PeerAddress> peerAddresses)
         {
-            Guard.NotNull(peerAddressManager, nameof(peerAddressManager));
+            Guard.NotNull(peerAddresses, nameof(peerAddresses));
 
-            this.peerAddressManager = peerAddressManager;
+            this.peerAddresses = peerAddresses;
         }
 
         /// <inheritdoc/>
         public PeerAddress SelectPeer()
         {
-            if (this.peerAddressManager.Peers.Tried().Any() == true &&
-                (this.peerAddressManager.Peers.New().Any() == false || new Random().Next(2) == 0))
-                return this.peerAddressManager.Peers.Tried().Random();
+            var tried = this.peerAddresses.Attempted().Concat(this.peerAddresses.Connected());
+            if (tried.Any() == true && (!this.peerAddresses.Fresh().Any() || new Random().Next(2) == 0))
+                return tried.Random();
 
-            if (this.peerAddressManager.Peers.New().Any() == true)
-                return this.peerAddressManager.Peers.New().Random();
+            if (this.peerAddresses.Fresh().Any())
+                return this.peerAddresses.Fresh().Random();
 
             return null;
         }
 
         /// <inheritdoc/>
         public IEnumerable<PeerAddress> SelectPeers()
         {
-            return this.peerAddressManager.Peers.Where(p => p.Value.Preferred).Select(p => p.Value);
+            return this.peerAddresses.Where(p => p.Value.Preferred).Select(p => p.Value);
```

Comment:
Ho do we determine `Preferred`?

Reply:
Hi @dangershony , these will all be removed/replaced in the last PR as well as properly documented. I'm just busy "setting" up for the last logic change 👍